Refactor summarizing container creation into container

heliocliu opened the pull request:
Fixes #4913 

Currently the summaryManager makes a request to the loader to create a summarizing container.  Refactor this logic upwards into a callback the container provides to the context (so the summaryManager doesn't need to touch the loader directly), and extract the loader calls so that the loader is never invoked directly when creating the container.  The loader will be removed as a property of the container (back-compat support aside) in a following change.

heliocliu committed:
pull summarizing container creation into container

anthony-murphy commented:
have you surveyed partner code to ensure they are not using the loader via the containerRuntime? if they are it may not be removable

heliocliu commented:
> 
> 
> have you surveyed partner code to ensure they are not using the loader via the containerRuntime? if they are it may not be removable

They're definitely using it, so if it's removed as a property it needs to be made available in some other way instead.  The question is probably if the use case (creating more containers) is considered fundamental enough that we don't want to make it an add-on feature.  In which case maybe instead of having a `createNextSummarizer` we just have a `createNewContainer` callback (which delegates this functionality into the original container), where creating new containers is "core" and anything else which is the loader is auxiliary?

vladsud commented:
I believe container runtime needs to be able to do the following:
1. Use its own loader to create new containers (with arbitrary URI and headers) - this is needed for "guest" components (as well as summarizer)
2. Be able to create new Loader instance (to control various features that apply to all containers in the loader, i.e. it's essentially creating new context for creating containers).

ChumpChief commented on packages/runtime/container-runtime/src/summaryManager.ts:
```diff
@@ -418,33 +423,44 @@ export class SummaryManager extends EventEmitter implements IDisposable {
+        } else {
+            const nextSummarizer = await this.context.createNextSummarizerFn(
+                this.context.deltaManager.lastSequenceNumber,
+                this.enableWorker ? "worker" : undefined,
```
Can we also extract the enableWorker flag all the way up to the Container?  Aside from simplifying the function signature would be good to consolidate all the opinionation on "how" to create a summarizer to one place I think.

heliocliu replied:
Newer iteration removes it from the function signature and keeps it in the headers.  For the future I think it can be added to the `IContainerLoadOptions` from your change, and the waiting on ops bit can be refactored into `Container.load` which would also let us isolate it's usage into the Loader (and we can deal with keeping it as a LoaderHeader option or whatever separately)

ChumpChief commented on packages/loader/container-definitions/src/runtime.ts:
```diff
@@ -127,6 +127,8 @@ export interface IContainerContext extends IDisposable {
     readonly submitSignalFn: (contents: any) => void;
     readonly snapshotFn: (message: string) => Promise<void>;
     readonly closeFn: (error?: ICriticalContainerError) => void;
+    readonly createNextSummarizerFn:
```
I think the callback is a good approach here, but wondering if you also consider event-based approaches (e.g. emit "becameSummarizer" or similar)?  Curious if you have thoughts there.

ChumpChief commented:
Re: the capabilities Vlad mentions -- this direction isn't attempting to forbid those capabilities, but as Helio mentions would rather just change the way they are accessed.  The goal is to avoid making the Loader ambiently available to the world but rather something that the author explicitly provides only if their scenario requires those capabilities.  This will help reduce coupling between our layers and restrict access to the scenarios that require them.

anthony-murphy commented:
I'm a bit torn here and would love description of the whole plan. To some degree this is layer breaking, as it's making the loader layer (Container) aware and coupled to a containerRuntime concept (summarization), and i'm not sure this is a coupling we want to introduce, as we our containerRuntime doesn't have to be the only containerRuntime. I think it would prefer if this was generalized more somehow, like a containerRuntime can get a clone or a copy of itself, regardless of if a loader is available, and then our specific containerRuntime could request it's summarizer off it. This also seems like it would more naturally support a future where we can summarize on a dirty containerRuntime, and don't need to get another, as we can just share with the user one

ChumpChief commented:
My perspective is that this coupling already exists functionally, it's just not explicitly discoverable.  The loader layer must respond in an expected way to a request from the ContainerRuntime (re: caching, URL shape, executionContext header, etc.) or else summarization breaks.  So although there's no import or explicit API call, you can't successfully write a Loader without being aware of the request semantic coming from the ContainerRuntime layer.

Also FWIW, I don't think we'll want to go in the direction of sharing a ContainerRuntime between user and summarizer.  It seems to me this would cause stalls in collaboration to the user as the summarization occurs (since we pause during summarization, and I don't currently see any good way around that).  More likely I think would be to share the driver layer on main thread but instantiate a separate ContainerRuntime on a web worker.

vladsud commented:
How about short meeting on that?

Here is how I see the world in future (and I think @curtisman has similar position):
I think the users of loader should not be able to change any settings of caching, pausing, forcing read-only state, etc. These should be the policies that apply to all containers in given loader, with no way to overwrite them. I.e. all apis to change them (Container.forceReadonly, paused state) should be removed from requests or container, and be provided as a policy (likely dynamic object) at loader creation time.

That's the only way to make things like caching, or pausing policy read-only work correctly for child containers that are created for "guest components". Host may chose (though policy) to say - I apply read-only state to ALL containers created in this loader, and that should impact guest components. Same is true to how we load guest components when it comes to pausing - right now that policy (expressed as a code in Bohemia repo) does not apply to guest components.

If client wants to escape these policies, it creates new loader. I think summarizer is in this bucket - it should not play games with passing various policies on request to existing loader (close on loss of connection, start unpaused, etc.) - it should express all of that by creating new loader and passing these policies at that moment.

heliocliu commented:
Looks like the plan is to go over the overall direction and goals for this and related work in Monday's design review.

To Vlad's point about loader policy, I think this change is generally moving toward that. The summarizer isn't using the existing loader's policies and doesn't try to override that loader's policies.  But instead of creating a new loader, the container exposes a method to create the summarizing copy.  Cleaning up loader policy headers is still a separate work item, along with having other scenarios in this situation create a new loader.

If we could just create a new runtime without needing to create a new container for the summarizer that would be cool 🤔

anthony-murphy commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1744,4 +1756,79 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+
+        const summarizerRequest: IRequest = {
+            headers,
+            url: "/_summarizer",
```
to me this is coupling the summarizer to the loader in a way it never was before, yes the container runtime depended on specific loader behaviors for requesting a container to summarize but this was similar to the loading flow of requesting any fluid object from any container. Ideally the loader(Container) doesn't know anything about summarziation, but provides the right api for the containerRuntime to implement it.

heliocliu replied:
At its core the API that's needed is one to create a new container copy with some options...  Changing this to create a new container and leaving the summarizer request part to the summaryManager would keep the container agnostic.  Though that leaves the options and runs counter to what Vlad was saying because it's likely overriding the loader options used to create the original container...

heliocliu committed:
Update summaryManager.ts

heliocliu committed:
Merge branch 'main' into RefactorSummarizer

heliocliu committed:
generalize callback

heliocliu committed:
Merge branch 'main' into RefactorSummarizer

heliocliu committed:
Update container.ts

heliocliu commented:
Still some other smaller things to address (like the scheduler) that should be up soon, but with the latest iteration I tried to address the coupling concerns while consolidating the stuff relating to loader policy into a smaller area to address in the separate loader header/option cleanup pr.

Besides the summarizer, there's also the scheduler that follows a similar pattern of requesting something on the original container, so the greatest common factor there is getting a "restricted" copy of the container, from which either the summarizer or scheduler can make a further request.  I changed the callback to be a more generic getContainerCopy function (specifically no caching and no reconnect - I think this is an okay starting point to address our internal scenarios).  It still uses the LoaderHeaders; this will be cleaned up to use the new create-a-new-loader paradigm in the separate loader header/option cleanup pr.  So we natively support a narrow container load API that fits our scenario and cannot be abused in the same way as the existing ambient loader for a "guest component" case or similar.

heliocliu committed:
propagate container creation fn to scheduler

could not find any coverage of the taskmanager, only the agentscheduler it holds.  so didn't add tests here because i'm not super sure how the task manager is supposed to be used

ChumpChief commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1730,6 +1732,15 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+            return this.loadContainerCopy(
+                this.loader,
+                this.loader.services.urlResolver,
+                this.originalRequest?.url,
```
I removed originalRequest in recent main, you'll want to merge it
```suggestion
                this.containerUrl,
```

ChumpChief commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1769,4 +1781,60 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+        loader: Loader,
+        urlResolver: IUrlResolver,
+        baseRequestUrl: string | undefined,
+        additionalHeaders: IRequestHeader,
```
Can this take an `IContainerLoadOptions` (or maybe some scoped subset of it) instead of an `IRequestHeader` perhaps?

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into RefactorSummarizer

heliocliu commented:
@vladsud I propagated the `loadContainerCopy` callback down to the data store level so the TaskManager can use it, but I wasn't able to find any actual usage of the TaskManager beyond using its internal AgentScheduler.   Do you know if there are any docs/usages I missed I can reference for some coverage tests?

ChumpChief commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1769,4 +1781,60 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+            containerRequest,
+            resolvedAsFluid);
+
+        const fromSequenceNumber = additionalHeaders[LoaderHeader.sequenceNumber] ?? -1;
```
And here probably either take the `sequenceNumber` directly as a param, or else return the `Container` as-loaded and let the caller wait on the `sequenceNumber` themselves.  I'm interested in trying to keep `LoaderHeader` out of `Container` :)

ChumpChief commented on packages/loader/container-definitions/src/runtime.ts:
```diff
@@ -120,6 +121,7 @@ export interface IContainerContext extends IDisposable {
     readonly submitFn: (type: MessageType, contents: any, batch: boolean, appData?: any) => number;
     readonly submitSignalFn: (contents: any) => void;
     readonly closeFn: (error?: ICriticalContainerError) => void;
+    readonly loadContainerCopyFn: (additionalHeaders: IRequestHeader) => Promise<IContainer>,
```
Let's have explicit parameters here to constrain what options can be provided.  With `IRequestHeader` it's not obvious what options might be passed, and what the implementation might be looking for.

ChumpChief reviewed (commented):
Looking pretty good -- I think my main thought is if we can more-or-less eliminate `IRequestHeader` as the means to pass the options around?

heliocliu committed:
split out request args

heliocliu committed:
lint errors

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into RefactorSummarizer

ChumpChief commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1842,4 +1862,65 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+        fromSequenceNumber?: number,
+        summarizingClient?: boolean,
+        // TODO #4912 eventually we may wish to spawn an execution context from which to run this
+        executionContext?: string,
```
Unused param, should this be removed?  Or are you planning to hook this up in a subsequent change?

heliocliu replied:
Yep planning to hook it up as part of the issue to do background summarization.  I guess since it's an optional anyway it's safer just to omit it for now in case we decide to change how it looks later

heliocliu committed:
remove unused executionContext param

anthony-murphy commented on packages/loader/container-loader/src/container.ts:
```diff
@@ -1799,6 +1801,21 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
+        const loadContainerCopyFn = async (
+            clientDetails?: IClientDetails,
+            fromSequenceNumber?: number,
+            summarizingClient?: boolean,
```
loader layer shouldn't be aware of summarization

heliocliu replied:
i dugged a dig and apparently this arg just doesn't matter.  it gets used by `odspDriverUrlResolver` to set a `summarizer` property on an `IOdspResolvedUrl`, but nothing ever reads it and the summarizer decides to do it's job based entirely on other things.

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -177,6 +179,14 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         return (await this.getInitialSnapshotDetails()).isRootDataStore;
     }
 
+    public get loadContainerCopyFn(): (
```
i don't think this should be public

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContext.ts:
```diff
@@ -177,6 +179,14 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         return (await this.getInitialSnapshotDetails()).isRootDataStore;
     }
 
+    public get loadContainerCopyFn(): (
```
why do we need this. i don't think it should exist. datastores shouldn't be creating container copies at will

heliocliu replied:
This was to pipe down the callback into `scheduler.ts` (above).  It's strictly less functionality than provided by the loader that gets piped down as well, though it's possible we don't even want this much by default...

I can remove it for now, but the scheduler needs to be accounted for somehow.  @ChumpChief has #4413 right now which involves potentially removing the container creation part of the scheduler, so depending on the decisions there it may not be an issue or we may potentially need to add this back

heliocliu replied:
Jk looked into this again, the paired ODSP url resolver and driver use it but the resolved url is passed around as an IResolvedUrl and cast to IOdspResolvedUrl in the driver...  The url resolver certainly doesn't care about that, but there isn't currently a good way to communicate that to the driver on the types used in the interface.  Going to add a client type -> header translation here as a temporary workaround until that's resolved

anthony-murphy commented on packages/runtime/runtime-definitions/src/dataStoreContext.ts:
```diff
@@ -267,6 +268,15 @@ IEventProvider<IFluidDataStoreContextEvents>, Partial<IProvideFluidDataStoreRegi
+    /**
+     * Load a restricted copy of the hosting container
+     */
+    readonly loadContainerCopyFn: (
```
we shouldn't expose this

heliocliu committed:
remove summarizer arg, data store callback

heliocliu committed:
Update container.ts
